### DISCUSSION TITLE
Question regarding truncating text labels in nodes

### DISCUSSION THREAD

**Discussion by david-pa** [eclipse-glsp/glsp#977] [Q&A]
I've posted here before since I am working on a project that uses GLSP to create a UI, and I have another question. Right now, there is currently a series of nodes with the following structure:

<img width="600" alt="image" src="https://user-images.githubusercontent.com/108750899/232914449-373cfb31-2ee7-428b-8123-bb17682b82c9.png">

Within the header, there is a label showing the title associated with the node, and there are also labels that make up the items in List A and List B, all of which are editable via the EditLabelUI extension.

Lately I've been working on solving issues related to scaling and maintaining reasonable proportions when the labels are exceptionally long. I was wondering if there is any possible way to show a label in a sort of truncated state. For example, a label that reads something like `aVeryLongNameForALabel` could be displayed as `aVery...ALabel` or some variation of that depending on the character count, but can still be viewed as `aVeryLongNameForALabel` when editing the label. If there isn't anything built-in to labels at this time, are there any concepts/properties currently part of labels that might be able to help implement something like this?

**Comment by david-pa** [comment]
@planger any assistance with this would be much appreciated!

**Comment by planger** [comment]
Unfortunately there is no native support for that in SVG and CSS. The property [`text-overflow`](https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow) isn't supported in most of the browsers as far as I know.

What you still can do is to use a custom `SLabelView` that checks for the length and whenever it would result in a larger width than some constant or computed max width, you could just truncate the label text directly before rendering it in SVG.

Note that there are two rendering passes:
1. Triggered by the `RequestBoundsAction` from the server, the client will render everything in a hidden div to obtain the size of all elements. During this pass, you don't have the sizes yet in the GModel/SModel. So you won't be able to truncate, but that's fine.
2. In the second pass, you should have the sizes of all elements in the GModel available and would be able to compute whether to truncate or not. However at this point the size of the label has already been determined without truncation. So the layouting algorithms that would depend on the label width will be wrong.

Thus, you can either truncate based on the string lentgh already in rendering pass one and this will result in correct width and layouting. Or you do it in the second pass and need to take into account -- e.g. in a custom layouter that takes into account that the label size is actually wrong (too long) if you know you'll truncate it in pass two.

**Reply by david-pa** [reply]
After looking at `SLabelView`, I see what you mean. That does seem like the best approach in terms of where to truncate, but would it be possible to store the original string somewhere so that the original string can appear when editing? Seems like `EditLabelUI` has a method called `applyTextContents` that has access to a label object with the type `(EditableLabel & SModelElement)`.

**Reply by planger** [reply]
Yes, you can store the original label in an argument of your `SLabel`. See also https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/base/argumentable.ts

**Reply by david-pa** [reply]
Sorry about the late reply, but @planger could you provide an example of how to store a value via `SLabel`/`SArgumentable`?

**Reply by planger** [reply]
On the server in the `GModelFactory` you can add an argument with `addArg()` of the GElement builder, e.g.
```typescript
GLabel.builder().text(task.name).id(`${task.id}_label`).addArg('originalLabel', 'text').build()
```

On the client, you can use `hasArguments(element)` to check if an element has arguments. E.g.
```typescript
if (hasArguments(element)) {
    const orignalLabel = element.args.originalLabel;
}
```
